rfcs/0051-emit-on-close-stream-executor.md

```diff
@@ -0,0 +1,158 @@
+  fn next(&mut self, table: &StateTable) -> Option<Row>;
+
+  /// get the minimum sort key in rows which is complete.
+  /// return None if there is no remaining complete row.
+  /// fill cache from the stateTable when cache miss.
+  /// will not consume the row in memory.
+  fn peek_sort_key(&self, table: &StateTable) -> Option<ScalarImpl>;
+
+  /// use the first column's watermark to refresh the SortBuffer
+  /// can trigger more rows complete which can be returned by `self.next`
+  fn accept_watermark(&mut self, watermark: Watermark);
+  
+  /// apply changes operations
+  fn insert(&mut self, row: Row);
+  fn delete(&mut self, row: Row);
+  fn apply_chunk(&mut self, c: Chunk);
+
+  /// recovery from the stateTable, just gets the row belonging to stateTable's vnode.
+  fn recovery(&mut self, table: &StateTable);
+}
+
+```
+
+The component does not take over so many things. There are still lots of things that should be done by the executor, such as
+
+- all the operations should be applied on the table and cache, and the executor should write in duplicate.
+- after getting the complete data, the executor must delete the useless data by itself (delete one by one/range delete with the last key). Notice that it is different with state clean and the data must be deleted consistently.
+
+And then we will see why and how it works in different situations.
+
+## Executors
+
+### Sort
+
+SortExecutor just has a executor which wraps a SortBuffer which can transform any stream into a append only stream ordered by a watermark column.
+
+- when stream chunk arrive
+  - write the chunk into stateTable
+  - write the chunk into SortBufferCache
+- when watermark arrive
+  - pass watermark into SortBufferCache
+- when barrier arrive
+  - drain complete rows from SortBufferCache, for each rows
+    - emit the row downstream
+    - delete the row in stateTable
+
+### SortAgg (with batch only agg calls)
+
+Under "EMIT ON WINDOW CLOSE" semantics, if there is watermark in group key, we can use SortExecutor for the input stream, and then we get a sorted stream which can be processed by "BatchSortAgg". Then, we can support all the aggregators which can run in batch mode.
+
+The SortAgg supports more kinds of aggregators under "EMIT ON WINDOW CLOSE", but should materialized all the input data, which could be worse than the normal streamGroupAgg. So we have another EMIT ON WINDOW CLOSE GroupAgg implementation as described below. I think the SortAgg is just to solve some aggregation functions which does not have a streaming version(hard to implement or with high cost)
+
+### GroupAgg
+
+If there is watermark in group key, the EMIT ON WINDOW CLOSE GroupAgg calculate the aggregation with the same logic with streamGroupAgg, But it only emit the complete part of the agg result. A simple idea is add a sortExecutor after a normal streamGroupAgg and we can find that the streamGroupAgg's result table and sortExecutor's stateTable exactly have the same schema and meaning.
```

Comment:
It seems to be unacceptable to me to use 2 caches in memory, which effectively means the memory is cut into 1/2, as well as the 2x IO cost of cache miss.

Reply:
I think the SortBufferCache here can be removed. It doesn't worth to maintain a full row copy in memory just for emitting when a `Window` event is received.

Reply:
Yes, what you say makes sense here, because every record will just be read one time and then will be deleted soon. The cache is meaningless here.